python/llm/example/CPU/Deepspeed-AutoTP/test_deepspeed.py

```diff
@@ -0,0 +1,123 @@
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+import os
+import torch
+from transformers import AutoModelForCausalLM, LlamaTokenizer, AutoTokenizer
+import deepspeed
+from bigdl.llm import optimize_model
+import torch
+import intel_extension_for_pytorch as ipex
+import time
+import argparse
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Predict Tokens using `generate()` API for Llama2 model')
+    parser.add_argument('--repo-id-or-model-path', type=str, default="meta-llama/Llama-2-7b-chat-hf",
+                        help='The huggingface repo id for the Llama2 (e.g. `meta-llama/Llama-2-7b-chat-hf` and `meta-llama/Llama-2-13b-chat-hf`) to be downloaded'
+                             ', or the path to the huggingface checkpoint folder')
+    parser.add_argument('--prompt', type=str, default="Once upon a time, there existed a little girl who liked to have adventures. She wanted to go to places and meet new people, and have fun",
+                        help='Prompt to infer')
+    parser.add_argument('--n-predict', type=int, default=32,
+                        help='Max tokens to predict')
+
+    args = parser.parse_args()
+    model_path = args.repo_id_or_model_path
+    local_rank = int(os.getenv("LOCAL_RANK", "0"))
+    world_size = int(os.getenv("WORLD_SIZE", "1"))
+    os.environ['RANK'] = os.getenv("LOCAL_RANK", "0") # for deepspeed
+
+    # Native Huggingface transformers loading
+    model = AutoModelForCausalLM.from_pretrained(
+        model_path,
+        device_map={"": "cpu"},
+        low_cpu_mem_usage=True,
+        torch_dtype=torch.float16,
+        trust_remote_code=True,
+        use_cache=True
+    )
+
+    # Parallelize model on deepspeed
+    model = deepspeed.init_inference(
+        model,
+        mp_size=world_size,
+        dtype=torch.float16,
+        replace_method="auto",
+    )
+
+    # Apply BigDL-LLM INT4 optimizations on transformers
+    model = optimize_model(model.module.to(f'cpu'), low_bit='sym_int4')
+    model = model.to(f'cpu:{local_rank}')
+    print(model)
+
+    # Load tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)
+
+    # Generate predicted tokens
+    with torch.inference_mode():
+        prompts = tuple([args.prompt] * 64)
+        # Batch tokenizing
+        input_ids = torch.tensor(
+                tokenizer.batch_encode_plus(prompts, truncation=True,
```

Comment:
It is found that deepspeed multi-instance on a single input suffers from a performance degeneration, therefore example behavior is changed to a batch input

Reply:
> It is found that deepspeed multi-instance on a single input suffers from a performance degeneration, therefore example behavior is changed to a batch input

We should focus on single request, not batch input